### ISSUE TITLE
Performance / code size metrics?

### ISSUE THREAD

**Issue by Xion** [lalrpop/lalrpop#360]
I have a potential use case for lalrpop that involves a large grammar of about 4000 textual alternatives (where the structure of a single alternative is very simple). I initially tried solving it using the `RegexSet` type from the `regex` crate, but the time taken compiling the regex set turned out to be significant, and the resulting DFA was hitting the default memory limits of the `regex` crate.

(For more context: I'm parsing the mod texts from item data that come from Path of Exile API. The spec of known mod texts patterns can be found here in JSON files [here](https://github.com/Xion/ezomyte/tree/f1b5b5785727e12cc354deef000fb6c6f15ff90c/data/mods), where the patterns contain one or more `#` signs which correspond to integer or float numbers. The expected lalrpop grammar would therefore be something like:

```rust
use ezomyte::model::item::mods::ModValues;
use ezomute::util::parse_number;

grammar;

pub mod_values: ModValues = {
    <"+"?> value " to maximum Life" => ModValues::one(<>),
    <"+"?> value " to Cold Resistance" => ModValues::one(<>),
    "Adds " value " to " value " Lightning Damage to Attacks" => ModValues::two(<>),
    // etc. (about 4000 more)
}

value: f64 = r"\+?(\d+(?:\.\d+)?)" => parse_number(<>).unwrap()
// parse_number tries both i64::from_str and f64:;from_str and unifies the result to be f64 
```

where the definition of `mod_values` nonterminal would be generated from the JSON files I've linked. The resulting parser would then receive input like `"+7% maximum Life"` -- coming from the player's item market -- and extract mod values (`7`, in this case)).

If I understand correctly, the way lalrpop works is to create all the parsing code at build time, as Rust functions and types, without any runtime preparation that's necessary before a parser can be used. This sounds great, as it would eliminate the costly compilation step that regexes suffer from.

I'm wondering, however, if it is known how a large grammar like this would:
* affect the size of the final binary
* perform adequately during actual parsing

So basically, my questions are: Has anyone tried using the `lalrpop` crate for something big like this before? What are the current known limits of the crate's performance? Have those limits been documented anywhere? Would it cope with the use case I've described?

**Comment by nikomatsakis** [comment]
You don't want to have big tokens like "to maximum ...". Instead, make those individual words each be a token, like so: "to" "maximum" etc. So your grammar might look like:

```
use ezomyte::model::item::mods::ModValues;
use ezomute::util::parse_number;

grammar;

pub mod_values: ModValues = {
    <"+"?> <value> "to" "maximum" "Life" => ModValues::one(<>),
    <"+"?> <value> "to" "Cold" "Resistance" => ModValues::one(<>),
    "Adds" <value> "to" "value" "Lightning" "Damage" "to" "Attacks" => ModValues::two(<>),
    // etc. (about 4000 more)
}

value: f64 = r"\+?(\d+(?:\.\d+)?)" => parse_number(<>).unwrap()
```

In any case, I think LALRPOP would be a fine choice for this. I'm not sure what kind of parse table size you will get though. Note that we *do* use `RegexSet` internally to identify words (e.g., "to", "Cold", etc), but I imagine that breaking into smaller tokens as I have done will help quite a bit there. Otherwise, you'd have to write your own tokenizer -- which is also reasonable. (Actually, this is a case where we would optimize too; hand-written tokenizers tend to separate out "keywords" like this and first match on a regex for any word, and then use a hashmap to find the keyword.)

**Comment by nikomatsakis** [comment]
> In any case, I think LALRPOP would be a fine choice for this

But I'm not sure, you have to try and see =) and there might be other ways to structure your grammar, as well, that would make it more (or less) efficient.

**Comment by nikomatsakis** [comment]
e.g., 

```
pub mod_values: ModValues = {
    <"+"?> <value> ("to" <Thing>)*  => ModValues::one(<>),
};

Thing = {
  "Cold" "Resistance", "Lightning" "Damage", ...
};
```

**Comment by Xion** [comment]
Thanks for the insights! I'll try a prototype and see how it performs.

**Comment by evbo** [comment]
The only problem I've seen with using small tokens is they are more likely to collide with other tokens. In general, the bigger the token the better chance of it being distinct (I think that's a true statement anyway...)

For instance, if you have a regex token for `"[a-zA-Z]+"` then `"to"` can't be a token but "to maximum" could be with the space.
